File: Bucket Sort/BucketSort.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

// Function to perform bucket sort on float values
void bucketSort(float *inputArr, int size) {
    // Create empty buckets
    vector<float> buckets[size];

    // Put array elements in different buckets
    for (int i = 0; i < size; i++) {
        int bucketIndex = size * inputArr[i]; // Index in the bucket
        buckets[bucketIndex].push_back(inputArr[i]);
    }

    // Sort individual buckets
    for (int i = 0; i < size; i++) {
        sort(buckets[i].begin(), buckets[i].end());
    }
    

    // Concatenate all buckets into the input array
    int index = 0;
    for (int i = 0; i < size; i++) {
        for (size_t j = 0; j < buckets[i].size(); j++) {
            inputArr[index++] = buckets[i][j];
        }
    }
}

// Function to perform bucket sort on integer values
void bucketSort(int *inputArr, int size) {
    // Find the maximum value in the array to determine the range of the buckets
    int max = inputArr[0];
    for (int i = 1; i < size; i++) {
        if (inputArr[i] > max) {
            max = inputArr[i];
        }
    }

    // Create the buckets
    int bucketCount = max / size + 1;
    vector<int> buckets[bucketCount];

    // Put array elements in different buckets
    for (int i = 0; i < size; i++) {
        int bucketIndex = inputArr[i] / size;
        buckets[bucketIndex].push_back(inputArr[i]);
    }

    // Sort individual buckets
    for (int i = 0; i < bucketCount; i++) {
        sort(buckets[i].begin(), buckets[i].end());
    }

    // Concatenate all buckets into the input array
    int index = 0;
    for (int i = 0; i < bucketCount; i++) {
        for (size_t j = 0; j < buckets[i].size(); j++) {
            inputArr[index++] = buckets[i][j];
        }
    }
}

int main() {
    char choice;
    cout << "Do you want to sort integers or floating-point numbers? (i/f): ";
    cin >> choice;

    if (choice == 'i') {
        int size;
        cout << "Enter size of the array: ";
        cin >> size;

        int *arr = new int[size];
        cout << "Enter " << size << " integer elements: ";
        for (int i = 0; i < size; i++) {
            cin >> arr[i];
        }

        cout << "Elements before Bucket Sort:" << endl;
        for (int i = 0; i < size; i++) {
            cout << arr[i] << " ";
        }
        cout << endl;

        bucketSort(arr, size);

        cout << "Elements after Bucket Sort:" << endl;
        for (int i = 0; i < size; i++) {
            cout << arr[i] << " ";
        }
        cout << endl;

        delete[] arr;
    } else if (choice == 'f') {
        int size;
        cout << "Enter size of the array: ";
        cin >> size;

        float *arr = new float[size];
        cout << "Enter " << size << " floating-point elements in the range [0, 1): ";
        for (int i = 0; i < size; i++) {
            cin >> arr[i];
        }

        cout << "Elements before Bucket Sort:" << endl;
        for (int i = 0; i < size; i++) {
            cout << arr[i] << " ";
        }
        cout << endl;

        bucketSort(arr, size);

        cout << "Elements after Bucket Sort:" << endl;
        for (int i = 0; i < size; i++) {
            cout << arr[i] << " ";
        }
        cout << endl;

        delete[] arr;
    } else {
        cout << "Invalid choice." << endl;
    }

    return 0;
}

File: Bucket Sort/BucketSort.py
def bucket_sort_float(input_arr):
    size = len(input_arr)
    # Create empty buckets
    buckets = [[] for _ in range(size)]

    # Put array elements in different buckets
    for i in range(size):
        bucket_index = int(size * input_arr[i])
        buckets[bucket_index].append(input_arr[i])

    # Sort individual buckets
    for i in range(size):
        buckets[i].sort()

    # Concatenate all buckets into the input array
    index = 0
    for i in range(size):
        for j in range(len(buckets[i])):
            input_arr[index] = buckets[i][j]
            index += 1

def bucket_sort_int(input_arr):
    size = len(input_arr)
    max_value = max(input_arr)
    
    # Create the buckets
    bucket_count = max_value // size + 1
    buckets = [[] for _ in range(bucket_count)]

    # Put array elements in different buckets
    for i in range(size):
        bucket_index = input_arr[i] // size
        buckets[bucket_index].append(input_arr[i])

    # Sort individual buckets
    for i in range(bucket_count):
        buckets[i].sort()

    # Concatenate all buckets into the input array
    index = 0
    for i in range(bucket_count):
        for j in range(len(buckets[i])):
            input_arr[index] = buckets[i][j]
            index += 1

def main():
    choice = input("Do you want to sort integers or floating-point numbers? (i/f): ")

    if choice == 'i':
        size = int(input("Enter size of the array: "))
        arr = [int(input(f"Enter element {i + 1}: ")) for i in range(size)]

        print("Elements before Bucket Sort:")
        print(" ".join(map(str, arr)))

        bucket_sort_int(arr)

        print("Elements after Bucket Sort:")
        print(" ".join(map(str, arr)))

    elif choice == 'f':
        size = int(input("Enter size of the array: "))
        arr = [float(input(f"Enter element {i + 1}: ")) for i in range(size)]

        print("Elements before Bucket Sort:")
        print(" ".join(map(str, arr)))

        bucket_sort_float(arr)

        print("Elements after Bucket Sort:")
        print(" ".join(map(str, arr)))

    else:
        print("Invalid choice.")

if __name__ == "__main__":
    main()
